.circleci/config.yml

```diff
@@ -0,0 +1,63 @@
+version: 2
+
+defaults: &defaults
+  working_directory: ~/repo
+  docker:
+    - image: circleci/node:10
+
+jobs:
+  setup:
+    <<: *defaults
+    steps:
+      - checkout
+      - restore_cache:
+          name: Restore node modules
+          keys:
+            - node_modules-{{ arch }}-{{ checksum "yarn.lock" }}
+      - run:
+          name: Install dependencies
+          command: yarn --pure-lockfile
+      - save_cache:
+          name: Save node modules
+          key: node_modules-{{ arch }}-{{ checksum "yarn.lock" }}
+          paths:
+            - node_modules
+
+  tests:
+    <<: *defaults
+    steps:
+      - checkout
+      - restore_cache:
+          name: Restore node modules
+          keys:
+            - node_modules-{{ arch }}-{{ checksum "yarn.lock" }}
+      - run:
+          name: Run Tests
+          command: yarn test
+
+  publish:
+    <<: *defaults
+    steps:
+      - checkout
+      - restore_cache:
+          name: Restore node modules
+          keys:
+            - node_modules-{{ arch }}-{{ checksum "yarn.lock" }}
+      - run:
+          name: Publish to NPM
+          command: yarn run release
+
+workflows:
+  version: 2
+  ci:
+    jobs:
+      - setup
+      - tests:
+          requires:
+            - setup
+      - publish:
+          requires:
+            - tests
+          filters:
+            branches:
+              only: master
```

Comment:
I would recommend changing the CircleCI config to the following. The reason being that you can't guarantee that the installed `node_modules` will work correctly between jobs. You are restoring based on the architecture, however, in the `tests` and `publish` steps there's not guarantee that they will be installed if the cache is not hit (different arch, or just CircleCI having a bad day).

The update I propose uses a CircleCI Orb which includes the correct logic for saving and restoring the yarn Cache.

```yaml
version: 2.1

orbs:
  rn: react-native-community/react-native@1.1.0

defaults: &defaults
  docker:
    - image: circleci/node:10

jobs:
  checkout_code:
    <<: *defaults
    steps:
      - checkout
      - persist_to_workspace:
          root: .
          paths: .
  
  test:
    <<: *defaults
    steps:
      - attach_workspace:
          at: .
      - rn/yarn_install
      - run:
        name: Run Tests
        command: yarn test
  
  publish:
    <<: *defaults
    steps:
      - attach_workspace:
          at: .
      - rn/yarn_install
      - run:
          name: Publish to NPM
          command: yarn run release

workflows:
  version: 2
  ci:
    jobs:
      - checkout_code
      - test:
          requires:
            - setup
      - publish:
          requires:
            - test
          filters:
            branches:
              only: master
```

Reply:
Awesome! 🙏